File: 2439-minimize-maximum-of-array/2439-minimize-maximum-of-array.cpp
class Solution {
public:

    int minimizeArrayValue(vector<int>& nums) {
        int n=nums.size() ;
        vector<long long int> pre(n,0) ;
        
        pre[0] = nums[0] ;
        for(int i=1 ; i<n ; i++)    pre[i] = pre[i-1] + nums[i] ;
        
        long long int ans=0 ;
        for(int i=0 ; i<n ; i++)
        {
            double d=i+1 ;
            double x = pre[i] / d ;
            long long int mx = ceil(x) ;
            ans = max( ans , mx ) ;
        }
        return ans ;
    }
};

File: 2439-minimize-maximum-of-array/2439-minimize-maximum-of-array.py
from math import ceil
from typing import List


class Solution:
    """
    Time:   O(n)
    Memory: O(1)
    """

    def minimizeArrayValue(self, nums: List[int]) -> int:
        cum_sum = maximum = 0

        for i, num in enumerate(nums, start=1):
            cum_sum += num
            maximum = max(ceil(cum_sum / i), maximum)

        return maximum